stats/opentelemetry/e2e_test.go

```diff
@@ -582,3 +613,362 @@ func pollForWantMetrics(ctx context.Context, t *testing.T, reader *metric.Manual
 
 	return fmt.Errorf("error waiting for metrics %v: %v", wantMetrics, ctx.Err())
 }
+
+// spanInformation is the information received about the span. This is a subset
+// of information that is important to verify that gRPC has knobs over, which
+// goes through a stable OpenTelemetry API with well-defined behavior. This keeps
+// the robustness of assertions over time.
+type spanInformation struct {
+	// SpanContext either gets pulled off the wire in certain cases server side
+	// or created.
+	sc         trace2.SpanContext
+	spanKind   string
+	name       string
+	events     []trace.Event
+	attributes []attribute.KeyValue
+}
+
+// TestClientCallSpanEvents verifies the events added to call spans
+// for a unary RPC, including events for gRPC status and name resolution
+// delays. It also verifies the same traceID is propagated across client
+// to server. It sets up a stub server with OpenTelemetry tracing enabled,
+// makes a unary RPC with gzip compression, and then asserts that the exported
+// spans contain the expected events and attributes.
+func (s) TestClientCallSpanEvents(t *testing.T) {
+	// Using defaultTraceOptions to set up OpenTelemetry with an in-memory exporter
+	traceOptions, spanExporter := defaultTraceOptions(t)
+
+	// Start the server with OpenTelemetry options
+	ss := setupStubServer(t, nil, traceOptions)
+	defer ss.Stop()
+
+	// Create a parent span for the client call
+	ctx, _ := otel.Tracer("grpc-open-telemetry").Start(context.Background(), "test-parent-span")
+	md, _ := metadata.FromOutgoingContext(ctx)
+	otel.GetTextMapPropagator().Inject(ctx, otelinternaltracing.NewCustomCarrier(ctx))
+	ctx = metadata.NewOutgoingContext(ctx, md)
+
+	grpc.NameResolutionDelayDuration = 0
+	// Make a unary RPC
+	if _, err := ss.Client.UnaryCall(
+		ctx,
+		&testpb.SimpleRequest{Payload: &testpb.Payload{Body: make([]byte, 10000)}},
+		grpc.UseCompressor(gzip.Name), // Deterministic compression.
+	); err != nil {
+		t.Fatalf("Unexpected error from UnaryCall: %v", err)
+	}
+
+	// Get the spans from the exporter
+	spans := spanExporter.GetSpans()
+	if got, want := len(spans), 3; got != want {
+		t.Fatalf("Got %d spans, want %d", got, want)
+	}
+
+	clientSpan := spans[2]
+	// Check that the client span has the correct status.
+	if got, want := clientSpan.Status.Code, otelcodes.Ok; got != want {
+		t.Errorf("Got client span status code %v, want %v", got, want)
+	}
+	// Check that the client has event for name resolution delay.
+	if got, want := clientSpan.Events[0].Name, "Delayed name resolution complete"; got != want {
+		t.Fatal("Client span didn't had event for delayed name resolution.")
+	}
+	// Check that same traceID is used in client and server.
+	if got, want := spans[0].SpanContext.TraceID(), spans[2].SpanContext.TraceID(); got != want {
```

Comment:
we should also check the attempt span id of client matches the span id of server SpanContext

Reply:
Added.